### ISSUE TITLE
BUG Check  "backreferences_dir" is str or None

### ISSUE THREAD

**Issue by banesullivan** [sphinx-gallery/sphinx-gallery#567]
All builds for [PyVista](https://github.com/pyvista/pyvista)'s docs are failing on the CIs with the new `v0.5.0` of `sphinx-gallery`. I confirmed this locally by upgrading `sphinx-gallery` and get the following error (from https://github.com/pyvista/pyvista/pull/451#issuecomment-555255085)

I'm not sure exactly where this error is coming from... seems to be only when using PyVsita's scraper - can you all help me track this down?

```
# Sphinx version: 2.2.1
# Python version: 3.7.3 (CPython)
# Docutils version: 0.15.2 release
# Jinja2 version: 2.10.3
# Last messages:

# Loaded extensions:
Traceback (most recent call last):
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/site-packages/sphinx/cmd/build.py", line 275, in build_main
    args.tags, args.verbosity, args.jobs, args.keep_going)
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/site-packages/sphinx/application.py", line 269, in __init__
    self._init_builder()
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/site-packages/sphinx/application.py", line 330, in _init_builder
    self.events.emit('builder-inited')
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/site-packages/sphinx/events.py", line 103, in emit
    results.append(callback(self.app, *args))
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/site-packages/sphinx_gallery/gen_gallery.py", line 284, in generate_gallery_rst
    seen_backrefs)
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/site-packages/sphinx_gallery/gen_rst.py", line 328, in generate_dir_rst
    fname, target_dir, src_dir, gallery_conf, seen_backrefs)
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/site-packages/sphinx_gallery/gen_rst.py", line 777, in generate_file_rst
    fname, intro)
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/site-packages/sphinx_gallery/backreferences.py", line 248, in _write_backreferences
    '%s.examples.new' % backref)
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/posixpath.py", line 94, in join
    genericpath._check_arg_types('join', a, *p)
  File "/Users/bane/anaconda3/envs/foo/lib/python3.7/genericpath.py", line 149, in _check_arg_types
    (funcname, s.__class__.__name__)) from None
TypeError: join() argument must be str or bytes, not 'bool'
```

**Comment by banesullivan** [comment]
All sites using PyVista's scraper are failing. Perhaps something changed with the Scraper API?

**Comment by lucyleeow** [comment]
I don't think the Scraper API changed. It is an unusual error. I'll work with @GuillaumeFavelier about this today (if he's in).

**Comment by GuillaumeFavelier** [comment]
I'm down, let's find out what's happening here.

**Comment by larsoner** [comment]
@GuillaumeFavelier @lucyleeow the fastest way to debug might be to start a Sphinx build in Python space. That way you can hit the error and post-mortem debug with `import pdb; pdb.pm()`. Something like this set of lines should hopefully allow you to replicate the error:

https://github.com/sphinx-gallery/sphinx-gallery/blob/master/sphinx_gallery/tests/test_full.py#L50-L56

**Comment by GuillaumeFavelier** [comment]
By using the `pdb` method explained above I was able to locate the issue around [those lines](https://github.com/sphinx-gallery/sphinx-gallery/blob/master/sphinx_gallery/backreferences.py#L246-L247). I inspected the value of `gallery_conf['backreferences_dir']` in post-mortem and it's `False`.

The error message makes sense since `op` tries to `join()` a path to a `bool` variable (the expected type is `str`).

At the [beginning of `_write_backreferences()`](https://github.com/sphinx-gallery/sphinx-gallery/blob/master/sphinx_gallery/backreferences.py#L246-L247), there is a test on `gallery_conf['backreferences_dir']` that returns if it's `None`. 

Maybe in the API, this test was using `False` and now it's using `None`?

**Comment by GuillaumeFavelier** [comment]
Or maybe the backreferences conf of PyVista changed somehow

**Comment by GuillaumeFavelier** [comment]
**TL;DR**:  `"backreferences_dir": None` in `pyvista/docs/conf.py` should solve this @banesullivan

**Comment by lucyleeow** [comment]
I think we can close this now. Good work @GuillaumeFavelier

**Comment by banesullivan** [comment]
Wow, thanks everyone for figuring this out so promptly!!

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by larsoner** [comment]
We should probably only allow `None` or `str`, so I'll reopen to remind us to add that check in our `gallery_conf` validation code.

**Comment by lucyleeow** [comment]
@larsoner I suggest altering `_write_backreferences()` ([here](https://github.com/sphinx-gallery/sphinx-gallery/blob/35fc739c5223aaa2a3516d1416d9c22c5cf30721/sphinx_gallery/backreferences.py#L239)) to check that `gallery_conf['backreferences_dir']` is `str` or `None` and throw an error if not OR if it isn't `str` or `None` change `gallery_conf['backreferences_dir']` to `None` and throw a warning. What do you think?

**Comment by lucyleeow** [comment]
Whoops, I realise the function to change should be `_complete_gallery_conf()` ([here](https://github.com/sphinx-gallery/sphinx-gallery/blob/35fc739c5223aaa2a3516d1416d9c22c5cf30721/sphinx_gallery/gen_gallery.py#L92))

**Comment by larsoner** [comment]
Agreed `_complete_gallery_conf` should check this

**Comment by lucyleeow** [comment]
Should it throw an error or change to `None` and throw a warning?

**Comment by larsoner** [comment]
I would say error, it's less likely to be missed/ignored
